### PR TITLE
ukf model

### PR DESCRIPTION
**Pull request by HongLouyemeng**
Implementation of a UKF model

### CODE REVIEW

File: pypose/module/UnscentedKalman.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    def compute_conv_mix(self,P,x_estimate,x_sigma,y_estimate,y_sigma):  #compute mix covariance
+
+        P_estimate = torch.zeros((P.shape),device= P.device,dtype=P.dtype)
+        for loop in range(1,self.loop_range):
+
+            e_x = x_sigma[loop - 1]- x_estimate
+            e_y = y_sigma[loop - 1]- y_estimate
+
+            P_estimate += torch.matmul(e_x , e_y.t()) * self.weight
```
Comment (wang-chen) [inline pypose/module/UnscentedKalman.py:53]: use tensor operation instead of loop

File: pypose/module/UnscentedKalman.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        for loop in range(1,self.loop_range):
+
+            e = sigma[loop -1 ] - estimate
+            P_estimate += self.weight * torch.matmul(e,e.t())
+
+        return P_estimate + noise
+
+
+    def matrix_conv(self,P,estimate,sigma,noise):  # Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches 14.49 Covariance matrix approximation， Is this correct ?
```
Comment (wang-chen) [inline pypose/module/UnscentedKalman.py:70]: need space after comma `,`

File: pypose/module/UnscentedKalman.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+        P_estimate = torch.zeros((P.shape), device=P.device, dtype=P.dtype)
+        for loop in range(1, self.loop_range):
+            print(estimate)
+            P_estimate += estimate @ P @estimate.t() *self.weight
+        return  P_estimate + noise
```
Comment (wang-chen) [inline pypose/module/UnscentedKalman.py:76]: use tensor instead of loop

File: pypose/module/UnscentedKalman.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        self.model.set_refpoint(state=x, input=u)
+        A, B = self.model.A, self.model.B
+        C, D = self.model.C, self.model.D
+        c1, c2 = self.model.c1, self.model.c2
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
+        x = bmv(A, x) + bmv(B, u) + c1
+        self.dim = x.shape[0]
+        self.weight = self.compute_weight()
+        self.loop_range = self.dim*2+1
```
Comment (wang-chen) [inline pypose/module/UnscentedKalman.py:104]: all operators need space `self.dim * 2 + 1` instead of `self.dim*2+1`

File: test/module/test_kalman.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ def observation(self, state, input, t):
             states[i+1], observ[i] = model(states[i] + w, inputs[i])
             estim[i+1], P[i+1] = ekf(estim[i], observ[i] + v, inputs[i], P[i], Q, R)
         error =  (states - estim).norm(dim=-1)
-
+        print(error)
```
Comment (wang-chen) [inline test/module/test_kalman.py:40]: remove this line

File: test/module/test_ukf.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+"""
+author: xiao haitao
+date:2022-12-19
+"""
+
+
+import pypose as pp
+import torch as torch
+class TestUKF:
```
Comment (wang-chen) [inline test/module/test_ukf.py:9]: add two space lines before class definition

File: pypose/module/__init__.py [hunk 7]
```diff
@@ -2,3 +2,4 @@
 from .kalman import EKF
 from .dynamics import System, LTI
 from .imu_preintegrator import IMUPreintegrator
+from .UnscentedKalman import UKF
```
Comment (wang-chen) [inline pypose/module/__init__.py:5]: rename `kalman.py to `ekf.py`, rename yours to `ukf.py`

**Review by wang-chen — CHANGES_REQUESTED** [review]

File: pypose/module/ukf.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
+import torch
+from torch import nn
+from ..basics import bmv
+from torch.linalg import pinv,matrix_power
+from torch.autograd import Function
+import numpy as np
+import scipy.linalg
+
+
+class MatrixSquareRoot(Function):
```
Comment (wang-chen) [inline pypose/module/ukf.py:10]: We don't suggest this implementation, as it only supports operation on CPU.

File: test/module/test_ukf.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+                super().__init__()
+
+            def state_transition(self, state, input, t=None):
+                return state.cos() + input
+
+            def observation(self, state, input, t):
+                return state.sin() + input
+
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        matrix_square_root_device = 'cpu'
```
Comment (wang-chen) [inline test/module/test_ukf.py:19]: don't suggestion setting the device independently.

**Review by wang-chen — CHANGES_REQUESTED** [review]
- Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
- The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
- If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.

File: examples/module/filter/bicycle_ukf.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        ax.add_artist(e)
+        e.set_facecolor("none")
+        e.set_edgecolor(color[i])
+    state_plot = ax.quiver(state[:-1, 0], state[:-1, 1],
+                           state[1:, 0] - state[:-1, 0], state[1:, 1] - state[:-1, 1],
+                           scale_units="xy", angles="xy", scale=1, color=color, label="True State")
+    est_plot, = ax.plot(est[:, 0], est[:, 1], '.-', label="Estimated State")
+    ax.legend(handler_map={est_plot: HandlerLine2D(numpoints=1)})
+    plt.title("UKF Example")
+    plt.show()
```
[the remark this answers is not in the document]
Comment (HongLouyemeng) [reply examples/module/filter/bicycle_ukf.py]: There is no standard matrix square root for pytorch,So we need to introduce new operations。

**Comment by HongLouyemeng** [comment]
> * Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
> * The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
> * If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.

**Comment by HongLouyemeng** [comment]
> > * Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
> > * The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
> > * If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.

**Comment by HongLouyemeng** [comment]
> > > * Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
> > > * The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
> > > * If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.

There is no standard matrix square root for pytorch,So we need to add new operations。

**Comment by wang-chen** [comment]
> > > > * Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
> > > > * The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
> > > > * If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.
> 
> There is no standard matrix square root for pytorch,So we need to add new operations。

Thank for confirming, then please expose those functions to users and add docs, examples, etc. Thank you so much!

**Comment by HongLouyemeng** [comment]
> > > > > * Please be careful to create the file `basics/utils.py`. It seems that it can be added to `basics/ops.py`.
> > > > > * The file `basics/utils.py` is totally unpolished, which doesn't meet the standards of `pypose` and cannot be merged.
> > > > > * If you believe that we have to provide new operators to implement UKF, it has to follow the convention of PyTorch, including documentation, naming policy, and implementation quality.
> > 
> > 
> > There is no standard matrix square root for pytorch,So we need to add new operations。
> 
> Thank for confirming, then please expose those functions to users and add docs, examples, etc. Thank you so much!

No problem .

File: docs/source/basics.rst [hunk 11]
```diff
@@ -45,3 +45,4 @@ Basics
     cummul_
     cumops
     cumops_
+    MatrixSquareRoot
```
Comment (wang-chen) [inline docs/source/basics.rst:48]: name convention doesn't follow, rename to `msqrt`

File: docs/source/ops.rst [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+.. currentmodule:: pypose
+.. autosummary::
+   :toctree: generated
+   :nosignatures:
+
+   :template: autosummary/class-no-inherit.rst
+   basics.bmv
+   basics.MatrixSquareRoot
+
```
Comment (wang-chen) [inline docs/source/ops.rst:12]: remove this file, we don't need a seperated `rst` file. both `bmv` and `msqrt` are in basics.rst

**Review by wang-chen — CHANGES_REQUESTED** [review]

File: pypose/basics/ops.py [hunk 13]
```diff
@@ -33,7 +33,7 @@ def bmv(input, vec, *, out=None):
     return torch.matmul(input, vec.unsqueeze(-1), out=out).squeeze(-1)
 
 
-class MatrixSquareRoot:
+class msqrt:
```
Comment (wang-chen) [inline pypose/basics/ops.py:36]: class definition should still use `Msqrt`. Define a new function called `msqrt` which internally calls `Msqrt.apply`, then the users can just call `msqrt(x)` like functions.

Comment (HongLouyemeng) [reply pypose/basics/ops.py:36]: Because there are many methods of matrix root mean square approximation, whether it is necessary to distinguish different methods by naming. Or use $msqrt$ uniformly?

File: pypose/module/ukf.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+        Return:
+            list of :obj:`Tensor`: posteriori state and covariance estimation
+        '''
+        # Upper cases are matrices, lower cases are vectors
+        self.model.set_refpoint(state=x, input=u)
+        A, B = self.model.A, self.model.B
+        C, D = self.model.C, self.model.D
+        c1, c2 = self.model.c1, self.model.c2
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
```
Comment (wang-chen) [inline pypose/module/ukf.py:262]: we don't use the linearized system for UKF. please remove.

File: pypose/module/ukf.py [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+        # x = bmv(A, x) + bmv(B, u) + c1
+
+        self.dim = x.shape[0]
+        self.weight = self.compute_weight()
+        self.loop_range = self.dim * 2 + 1
+
+        # compute sigma point,mean,covariance
+        x_sigma, y_sigma = self.compute_sigma(x, u, P, C, D, c2)
+        x_sigma = bmv(A, x_sigma) + bmv(B, u) + c1
```
Comment (wang-chen) [inline pypose/module/ukf.py:271]: UKF uses `self.model` for system transition, not linearized A,B,C,D,c1,c2

File: pypose/module/ukf.py [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+        index = torch.ones(self.dim, self.dim, device=P.device, dtype=P.dtype)
+
+        index_gather = torch.as_tensor(index * index_weight, dtype=torch.int64, device=P.device).repeat(
+            index_repeat).unsqueeze(1).reshape(2 * self.dim, 1, self.dim)
+
+        # compute root of P
+        repeat_dim = tuple(torch.cat([torch.tensor([self.dim * 2]), torch.tensor(P.shape)]).numpy())  # repeat dim
+        p_expand = P.expand(repeat_dim)
+        np_expand = self.dim * p_expand  # calculate np
+        np_expand_sqrt = msqrt(np_expand)  # square root of np
```
Comment (wang-chen) [inline pypose/module/ukf.py:194]: `P` is always positive definite, thus we may be able to use [Cholesky decomposition](https://pytorch.org/docs/stable/generated/torch.linalg.cholesky.html) for matrix square root, which could be faster. You may want to test for speed comparison.

**Review by wang-chen — CHANGES_REQUESTED** [review]
general comments:

- make sure all lines are less than 97 characters.

File: pypose/module/ukf.py (hunk 14 above)
Comment (HongLouyemeng) [reply pypose/module/ukf.py]: Should we keep $Q$ and $R$ to add noise for covariance ? $$\mathbf{P}_{y} = \frac{1}{2n} \sum\_{i=1}^{2n} \left ( \mathbf{y}^{(i)} - \hat{\mathbf{y}} \right ) \left ( \mathbf{y}^{(i)} -\hat{\mathbf{y} } \right ) ^T+\mathbf{R}$$

File: pypose/module/ukf.py (hunk 15 above)
Comment (HongLouyemeng) [reply pypose/module/ukf.py]: That's right,  When using a linearized A,B,C,D,c1,c2, the ukf model cannot process 100 dimensions in test_ukf.py.

File: pypose/module/ukf.py (hunk 16 above)
Comment (HongLouyemeng) [reply pypose/module/ukf.py:220]: https://github.com/HongLouyemeng/pypose/blob/e28bb3a81378046c6cd46e642a41f8daab4b15ed/pypose/basics/ops.py#L214-L273, Cholesky in pytorch will make the ukf unstable, I have found a method, but its covariance will make the eigenvalue negative in **Bicycle Model**,but it will not occur based on **contour_integral** or **pade**.

Comment (HongLouyemeng) [reply pypose/module/ukf.py:220]: Reference:
  [1][https://iopscience.iop.org/article/10.1209/0295-5075/78/12001](url)
  [2][http://eprints.maths.manchester.ac.uk/1136/1/hht08.pdf](url)

Comment (HongLouyemeng) [reply pypose/module/ukf.py]: Cholesky decomposition  is faster than Padé approximant.

File: pypose/basics/ops.py [hunk 17]
```diff
@@ -1,5 +1,8 @@
 import torch
 from .. import LieTensor
+import numpy as np
+import mpmath as mpm
```
Comment (wang-chen) [inline pypose/basics/ops.py:4]: only use `torch` library

Comment (HongLouyemeng) [reply pypose/basics/ops.py:4]: ok

Comment (HongLouyemeng) [reply pypose/basics/ops.py:4]: Taylor approximation requires faster function to calculate higher order derivatives.

Comment (wang-chen) [reply pypose/basics/ops.py:4]: Are you able to use analytical term for each order?

Comment (wang-chen) [reply pypose/basics/ops.py:4]: If you prefer not to implement it now (leave it for the future), then you may remove this function and the UKF is ready to merge I think.

Comment (HongLouyemeng) [reply pypose/basics/ops.py]: > If you prefer not to implement it now (leave it for the future), then you may remove this function and the UKF is ready to merge I think.

https://github.com/pytorch/pytorch/issues/93270https://github.com/pytorch/pytorch/issues/93270

**Review by wang-chen — APPROVED** [review]
